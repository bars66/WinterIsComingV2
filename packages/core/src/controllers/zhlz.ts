import {AbstractController} from './abstract';
import type {Context} from '../context';
import type {Bridge} from '../bridges/bridge';
import {SqlSettings} from '../utilities/sqlSettings';
import {retry} from '../utilities/retry';

export class Zhlz extends AbstractController {
  static STATUS_DEVICE_ERROR = 'device_error';
  static DEFAULT_SETTINGS = {
    minPosition: [0, 0, 0],
    maxPosition: [3520, 3520, 3520],
    currentPosition: [0, 0, 0],
    inverted: [true, false, false],
  };

  private bridge: Bridge;
  private clientId: number;
  private settings: SqlSettings<typeof Zhlz.DEFAULT_SETTINGS>;

  constructor(id: string, context: Context, bridge: Bridge, clientId: number) {
    super(id, context, 'Zhlz controller');

    this.bridge = bridge;
    this.clientId = clientId;
    this.settings = new SqlSettings<typeof Zhlz.DEFAULT_SETTINGS>(
      context,
      `${this.name}_${this.id}`,
      Zhlz.DEFAULT_SETTINGS
    );
  }

  protected async fetchStatus(): Promise<string> {
    const answer = await this.bridge.execute({
      clientId: this.clientId,
      retry: 1,

      command: {
        cmd: 'readHoldingRegisters',
        dataAddress: 0,
        length: 3,
      },
    });

    this.logger.trace({answer}, 'Get status from device');

    if (answer.error) {
      this.logger.error({error: answer.error}, 'Error from device');

      return Zhlz.STATUS_DEVICE_ERROR;
    }

    if (!!answer.answer) {
      return Zhlz.STATUS_AVAILABLE;
    }

    return Zhlz.STATUS_UNAVAILABLE;
  }

  getCmdValue(
    settings: typeof Zhlz.DEFAULT_SETTINGS,
    newPositions: Array<number>
  ): {cmd: Array<number>; currentPositions: Array<number>} {
    let cmd = [];
    let currentPositions = [];

    if (newPositions.length !== settings.currentPosition.length) {
      throw new Error('Длинны не совпадают');
    }

    for (let i: number = 0; i !== newPositions.length; ++i) {
      const newPosition = newPositions[i];

      const isInverted = settings.inverted[i];
      const minPosition = settings.minPosition[i];
      const maxPosition = settings.maxPosition[i];
      const currentPosition = settings.currentPosition[i];

      if (newPosition == undefined || newPosition === currentPosition) {
        cmd.push(0);
        currentPositions.push(currentPosition);
        continue;
      }

      const position: number = Math.min(Math.max(newPosition, minPosition), maxPosition);
      const delta = position - currentPosition;

      cmd.push(delta * (isInverted ? -1 : 1));
      currentPositions.push(position);
    }

    return {cmd, currentPositions};
  }

  async setPositions(newPositions: Array<number>): Promise<void> {
    const settings = await this.settings.getSettings();
    const newValues = this.getCmdValue(settings.settings, newPositions);

    try {
      await this.bridge.execute({
        clientId: 10,
        retry: -1,

        command: {
          cmd: 'writeRegisters',
          dataAddress: 0,
          values: newValues.cmd.map((value) => (value === 0 ? 0 : value + (1 << 15))),
        },
      });
    } catch (e) {
      this.logger.error({error: e, settings, newPositions}, `Can't change position`);
      return;
    }

    settings.settings.currentPosition = newValues.currentPositions;

    try {
      const fn = async () => {
        await this.settings.write(settings.settings, settings.updateId, true);
      };

      await retry(fn, {attempts: 5});
    } catch (e) {
      this.logger.fatal({error: e, settings, newPositions}, `Can't change settings`);
    }
  }

  private async changeAllPosition(newValue: number): Promise<void> {
    const settings = await this.settings.getSettings();
    await this.setPositions(settings.settings.currentPosition.map((_) => newValue));
  }

  public executeAction(action: string, params?: string): Promise<void> {
    switch (action) {
      case 'open': {
        return this.changeAllPosition(+Infinity);
      }

      case 'close': {
        return this.changeAllPosition(-Infinity);
      }
    }

    throw new Error('Unknown action ' + action);
  }
}
