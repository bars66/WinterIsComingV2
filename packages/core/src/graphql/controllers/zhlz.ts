import {GraphQLInt, GraphQLObjectType, GraphQLList, GraphQLString} from 'graphql';
import {Zhlz} from '../../controllers/zhlz';
import {Controller, fields} from './controller';

export const ZhlzType = new GraphQLObjectType({
  name: 'Zhlz',
  interfaces: () => [Controller],
  fields: () => ({
    ...fields,
    position: {
      type: new GraphQLList(GraphQLInt),
      resolve(controller: Zhlz): Promise<Array<number>> {
        return controller.getPosition();
      },
    },
    motorsStatus: {
      type: new GraphQLList(GraphQLString),
      resolve(controller: Zhlz): Array<string> {
        return controller.getMotorsStatus();
      },
    },
  }),
});
