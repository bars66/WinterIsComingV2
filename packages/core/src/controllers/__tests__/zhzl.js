import {Zhlz} from '../zhlz';

jest.unmock('../zhlz');
jest.unmock('../../utilities/retry');

// По сути любое кручение шаговым двигателем возможно, не только жаллюзи
describe('Привод управления жаллюзи', () => {
  let sut;

  beforeEach(() => {
    sut = new Zhlz();
  });

  describe('Высчитывание положения', () => {
    let settings;

    beforeEach(() => {
      settings = {
        currentPosition: [1000],
        minPosition: [0],
        maxPosition: [3500],
        inverted: [false],
      };
    });

    it('Крутим вверх', () => {
      expect(sut.getCmdValue(settings, [2000])).toEqual({cmd: [1000], currentPositions: [2000]});
    });

    it('Крутим вверх, inverted', () => {
      settings.inverted[0] = true;
      expect(sut.getCmdValue(settings, [2000])).toEqual({cmd: [-1000], currentPositions: [2000]});
    });

    it('Крутим вниз', () => {
      expect(sut.getCmdValue(settings, [750])).toEqual({cmd: [-250], currentPositions: [750]});
    });

    it('Крутим вниз, inverted', () => {
      settings.inverted[0] = true;
      expect(sut.getCmdValue(settings, [750])).toEqual({cmd: [250], currentPositions: [750]});
    });

    it('Стоим не месте', () => {
      expect(sut.getCmdValue(settings, [undefined])).toEqual({cmd: [0], currentPositions: [1000]});
    });

    it('Границы корректно обрабатываются', () => {
      expect(sut.getCmdValue(settings, [3500])).toEqual({cmd: [2500], currentPositions: [3500]});
      expect(sut.getCmdValue(settings, [4000])).toEqual({cmd: [2500], currentPositions: [3500]});
      expect(sut.getCmdValue(settings, [+Infinity])).toEqual({
        cmd: [2500],
        currentPositions: [3500],
      });

      expect(sut.getCmdValue(settings, [0])).toEqual({cmd: [-1000], currentPositions: [0]});
      expect(sut.getCmdValue(settings, [-1000])).toEqual({cmd: [-1000], currentPositions: [0]});
      expect(sut.getCmdValue(settings, [-Infinity])).toEqual({cmd: [-1000], currentPositions: [0]});
    });

    it('Много за раз', () => {
      settings = {
        currentPosition: [1000, 500, 300, 1000],
        minPosition: [0, 0, 0, 0],
        maxPosition: [3500, 3400, 3100, 3200],
        inverted: [false, true, false],
      };

      expect(sut.getCmdValue(settings, [+Infinity, +Infinity, undefined, 300])).toEqual({
        cmd: [2500, -2900, 0, -700],
        currentPositions: [3500, 3400, 300, 300],
      });
    });

    it('Если не совпдает длинна с настройками, то кинет ошибку', () => {
      settings = {
        currentPosition: [1000, 500, 300, 1000],
        minPosition: [0, 0, 0, 0],
        maxPosition: [3500, 3400, 3100, 3200],
        inverted: [false, true, false],
      };

      function test() {
        sut.getCmdValue(settings, [+Infinity]);
      }

      expect(test).toThrowErrorMatchingInlineSnapshot(`"Длинны не совпадают"`);
    });
  });

  describe('setPositions', () => {
    let execute;
    let settings;
    let settingsWrite;

    beforeEach(() => {
      settings = {
        settings: {
          currentPosition: [1000, 500, 300, 1000],
          minPosition: [0, 0, 0, 0],
          maxPosition: [3500, 3400, 3100, 3200],
          inverted: [false, true, false],
        },
      };

      execute = jest.fn(async () => {});
      settingsWrite = jest.fn(async () => {});

      sut.settings.write = settingsWrite;
      sut.settings.getSettings = jest.fn(async () => settings);
      sut.bridge = {execute};
      sut.logger = {error: jest.fn(), fatal: jest.fn()};
    });

    it('Отправит новые значения', async () => {
      await sut.setPositions([+Infinity, +Infinity, undefined, 300]);

      expect(execute).toMatchSnapshot();
      expect(settingsWrite).toMatchSnapshot();
    });

    it('Ошибка от устройства - не будем менять настройки', async () => {
      sut.bridge.execute = jest.fn(async () => {
        throw new Error('Bridge error');
      });

      await sut.setPositions([+Infinity, +Infinity, undefined, 300]);

      expect(settingsWrite).not.toBeCalled();
      expect(sut.logger).toMatchSnapshot();
    });

    it('Ошибка при записи в базу', async () => {
      sut.settings.write = jest.fn(async () => {
        throw new Error('Write error');
      });

      await sut.setPositions([+Infinity, +Infinity, undefined, 300]);

      expect(sut.settings.write).toHaveBeenCalledTimes(5);
      expect(sut.logger).toMatchSnapshot();
    });
  });
});
