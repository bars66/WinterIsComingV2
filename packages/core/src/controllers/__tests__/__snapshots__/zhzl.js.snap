// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Привод управления жаллюзи setPositions Отправит новые значения 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "clientId": 10,
        "command": Object {
          "cmd": "writeRegisters",
          "dataAddress": 0,
          "values": Array [
            35268,
            29868,
            0,
            32068,
          ],
        },
        "retry": -1,
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;

exports[`Привод управления жаллюзи setPositions Отправит новые значения 2`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "currentPosition": Array [
          3500,
          3400,
          300,
          300,
        ],
        "inverted": Array [
          false,
          true,
          false,
        ],
        "maxPosition": Array [
          3500,
          3400,
          3100,
          3200,
        ],
        "minPosition": Array [
          0,
          0,
          0,
          0,
        ],
      },
      undefined,
      true,
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;

exports[`Привод управления жаллюзи setPositions Ошибка от устройства - не будем менять настройки 1`] = `
Object {
  "error": [MockFunction] {
    "calls": Array [
      Array [
        Object {
          "error": [Error: Bridge error],
          "newPositions": Array [
            Infinity,
            Infinity,
            undefined,
            300,
          ],
          "settings": Object {
            "settings": Object {
              "currentPosition": Array [
                1000,
                500,
                300,
                1000,
              ],
              "inverted": Array [
                false,
                true,
                false,
              ],
              "maxPosition": Array [
                3500,
                3400,
                3100,
                3200,
              ],
              "minPosition": Array [
                0,
                0,
                0,
                0,
              ],
            },
          },
        },
        "Can't change position",
      ],
    ],
    "results": Array [
      Object {
        "type": "return",
        "value": undefined,
      },
    ],
  },
  "fatal": [MockFunction],
}
`;

exports[`Привод управления жаллюзи setPositions Ошибка при записи в базу 1`] = `
Object {
  "error": [MockFunction],
  "fatal": [MockFunction] {
    "calls": Array [
      Array [
        Object {
          "error": [Error: Write error],
          "newPositions": Array [
            Infinity,
            Infinity,
            undefined,
            300,
          ],
          "settings": Object {
            "settings": Object {
              "currentPosition": Array [
                3500,
                3400,
                300,
                300,
              ],
              "inverted": Array [
                false,
                true,
                false,
              ],
              "maxPosition": Array [
                3500,
                3400,
                3100,
                3200,
              ],
              "minPosition": Array [
                0,
                0,
                0,
                0,
              ],
            },
          },
        },
        "Can't change settings",
      ],
    ],
    "results": Array [
      Object {
        "type": "return",
        "value": undefined,
      },
    ],
  },
}
`;
